File: FirstandLast-34.py
# Time Complexity : O(logn)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : There's a small change in left and right bianry search while checking for nums[mid] > target


# Your code here along with comments explaining your approach 

class Solution(object):
    def leftMost(self, nums, target):
        low = 0
        high = len(nums) - 1

        while low <= high:
            mid = low + (high-low)//2
            if nums[mid] == target and (mid == 0 or nums[mid-1] != target):
                return mid
            elif nums[mid] >= target:
                high = mid - 1
            else:
                low = mid + 1
        return -1
    
    def rightMost(self, nums, target):
        low = 0
        high = len(nums) - 1

        while low <= high:
            mid = low + (high-low)//2
            if nums[mid] == target and (mid == len(nums) - 1 or nums[mid+1] != target):
                return mid
            elif nums[mid] >= target:
                high = mid - 1
            else:
                low = mid + 1
        return -1

    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        left = self.leftMost(nums, target)
        if left == -1:
            return [-1,-1]
        right = self.rightMost(nums, target)
        return [left,right]

s = Solution()
print s.searchRange([5,7,7,8,8,10], 10)

File: firstAndLast34.java
package nov18th2020;

// S30 Big N Problem #7 {Medium }
// Source Link: https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
// https://github.com/super30admin/Binary-Search-2
public class firstAndLast34 {
	private int binaryfirst(int[] nums, int target) {
		int low = 0, high = nums.length - 1;
		while (low <= high) {
			int mid = low + (high - low) / 2;
			if (nums[mid] == target) {
				if (nums[mid] == target) {
					if (mid == low || nums[mid] > nums[mid - 1]) {
						return mid;
					} else {
						high = mid - 1;
					}
				}
			} else if (target > nums[mid]) {
				low = mid + 1;
			} else {
				high = mid - 1;
			}
		}
		return -1;
	}

	private int binaryLast(int[] nums, int target) {
		int low = 0, high = nums.length - 1;
		while (low <= high) {
			int mid = low + (high - low) / 2;
			if (nums[mid] == target) {
				if (nums[mid] == target) {
					if (mid == high || nums[mid] < nums[mid + 1]) {
						return mid;
					} else {
						low = mid + 1;
					}
				}
			} else if (target > nums[mid]) {
				low = mid + 1;
			} else {
				high = mid - 1;
			}
		}
		return -1;
	}

	public int[] searchRange(int[] nums, int target) {
		if (nums == null || nums.length == 0)
			return new int[] { -1, -1 };
		int firstOccurence = binaryfirst(nums, target);
		int lastOccurence = binaryLast(nums, target);
		return new int[] { firstOccurence, lastOccurence };

	}
}

//. Time Complexity : O(log n)
// Space Complexity : O (1)

